### ISSUE TITLE
XEP-0363: HTTP File Upload 

### ISSUE THREAD

**Issue by ghost** [conversejs/converse.js#161]
Hello.

It probably hasn't been implemented from what I've seen, but is there/could there be a way of handling file upload/transfer from within the chat?

For example, I'd like to select (or drag and drop) a file from my drive in a conversation (either room or private). At that point, probably Converse.js would not handle the file itself, but rather offer a hook when such an event occurs. I have no experience with such events actually.
In the end, once the upload is considered over, an URL would be sent as a message, so chat users can access the file right away.

Thanks in advance.

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/2304598-http-file-upload?utm_campaign=plugin&utm_content=tracker%2F194169&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F194169&utm_medium=issues&utm_source=github).

<bountysource-plugin>

**Comment by kruks23** [comment]
Hello,

What about Jingle File Transfer: http://xmpp.org/extensions/xep-0234.html?

Conversations, gajim and swift implement that.

Thank you.

**Comment by jcbrand** [comment]
@kruks23 Thanks for the info. I haven't yet decided upon a standard, but good to know that XEP-0234 is implemented by those projects.

**Comment by chihanlin** [comment]
There are some WebRTC js available. For example, jsSIP, SIP.js, if converse.js can aware any one of these two, converse.js + (jsSIP or SIP.js) could be a great "Web based" UC client. those js do support file transfer, plus video, voice support,  they connect to the sip based IP-PBX. It will be a powerful integration client.

**Comment by chihanlin** [comment]
The file transfer probably can use http post... if it make sense... The user can upload the file, and the http server reply back the file URL to the sender when the file upload successfully , and the file sender then send out the URL to received side by using xmpp message. When the Received side get the message, the received side can show the URL and allow the user to click the URL to download the file from the http server. It is quite simple in this way.

**Comment by kruks23** [comment]
I think now is better to use Http File Upload (#466).

**Comment by nodiscc** [comment]
Note that this requires a third-party file storage service:

> The actual upload of the file happens via HTTP-PUT and is out of scope of this document

Whereas https://xmpp.org/extensions/xep-0096.html requires no additional server (if I understand correctly)

**Comment by kruks23** [comment]
Http FIle Upload requires the same server that you use to send your messages via xmpp. Servers like ejabberd or prosody already support xep-0363 Http file upload. 
it is something like xep-0313 Message archive management, messages are stored in the xmpp server.

**Comment by devurandom** [comment]
I created #593 for [XEP-0234](http://xmpp.org/extensions/xep-0234.html) (Jingle File Transfer) support. Could this issue please be renamed to "XEP-0363 - HTTP File Upload" (and add a link to [XEP-0363](http://xmpp.org/extensions/xep-0363.html) to clarify the destinction?

@nodiscc I think many clients are migrating away from [XEP-0096](https://xmpp.org/extensions/xep-0096.html) (SI File Transfer) and new clients are not implementing it, but are rather going straight for XEP-0234. The reason is afaik that [XEP-0095](http://xmpp.org/extensions/xep-0095.html) (Stream Initiation), on which SI File Transfer is based, does not deal with firewalls and NAT. This is handled by [XEP-0166](http://xmpp.org/extensions/xep-0166.html) (Jingle), on which Jingle File Transfer is based. Details can be found in the Requirements section of the Jingle File Transfer specification.

**Comment by jkufner** [comment]
HTTP storage is provided by Jabber server (ejabberd and prosody provides this). So to implement this feature, two tasks to be done: 1. Negotiate upload slot via XMPP, 2. perform HTTP PUT on the negotiated URL. That's it. Once upload is completed, URL needs to be sent via chat, but that is an ordinary message. To receive the uploaded file, nothing has to be done, the URL will appear in chat and that is completely fine.

HTTP upload is the most reliable file sharing method Jabber has. There is no need to implement anything else, since nothing else is actually supported widely enough and even when it is supported clients must understand each other (XEP versions) and must be able to connect to each other (NAT, firewall). It never really worked for me. On the other side, HTTP upload works well in MUC and with transports (ICQ, Facebook), since the only thing to send via chat is plain URL.

**Comment by devurandom** [comment]
> HTTP storage is provided by Jabber server [...]
> There is no need to implement anything else, since nothing else is actually supported widely enough [...]

In my experience, XMPP servers (instances, not the software) which support modern XEPs are extremely rare. So rare in fact, that I am again running an own instance, because none of the available (free or part of a paid email contract) ones support any useful (mobile) extensions.

Thus I would argue that a file transfer implementation which is independent of the server is still very much needed.

**Comment by jkufner** [comment]
@devurandom  As I understand purpose of Converse.js, it is meant for integration into another application, therefore you run your own infrastructure more or less anyway.

**Comment by tbeitter** [comment]
I'd love to see http_uploade here, too. +1

**Comment by devurandom** [comment]
This would likely also need support [XEP-0370](https://xmpp.org/extensions/xep-0370.html) to integrate nicely with other clients. I created issue #966 to track implementation of that XEP.

**Comment by happy-dev** [comment]
@jcbrand : Would you set me on track for this ? :) 
I could give it a try

**Comment by jcbrand** [comment]
@happy-dev You'll need to write a plugin for converse.js

The most up to date docs for plugin development are here:
https://github.com/jcbrand/converse.js/blob/master/docs/source/plugin_development.rst

Once the next release is made, those docs will also be on https://conversejs.org

**Comment by happy-dev** [comment]
Thank you very much !
I share this with the rest of the team and get back to you

**Comment by happy-dev** [comment]
Team OK for it. 
My idea of the implementation would be : 
 - configure some cloud storage solution
 - upload the file to it
 - post back a link on the chat when the file is done uploading on the cloud storage

Would you have any recommendation to tackle this ?
Do you believe this is an acceptable way to handle the feature ? 
Is there any existing function or XEP I should take into account before getting started ? 

Many thanks for your inputs

**Comment by jcbrand** [comment]
@happy-dev Sounds great :)

Generally how I go about adding new features/plugins, is that I first write the plugin outline, and make sure it's loaded together with all the other plugins.

That requires a few steps, such as updating the *src/config.js* file, adding the file to the *src* directory and whitelisting it in converse-core.

Most of that is covered in the plugin documentation mentioned above. The only thing not mentioned that I can think of now is the fact that you also need to set the plugin path in *src/config.js*.

Then, once you have the plugin registered and loading when you load converse.js, you can start adding the functionality via test-driven development.

I don't always do test-driven development, or let's say at least, I don't always start with tests. But with the XMPP protocol it works well because you can directly test the XML stanzas in the relevant XEP and then check that converse.js does the right thing.

So, once the plugin outline is available, I'd study the relevant [XEP-0363](https://xmpp.org/extensions/xep-0363.html) and start mocking the XML traffic.

This is fairly easy to do. Take a look at *spec/protocol.js* for tests which test the XMPP protocol. Similar things are done in *spec/chatbox.js* and *spec/chatroom.js*.

Basically you can simulate incoming XML stanzas, write your plugin to send the right outgoing stanzas and then test that this all conforms to the XEP.

This way you can write a plugin without needing an XMPP server at all. At least initially, obviously later you'll want to test with a real server.

**Comment by happy-dev** [comment]
Thank you so much. I believe you just set the TODO list for this weekend ! 
See you on Monday ;)

**Comment by jcbrand** [comment]
@happy-dev You might want to hang out in the discuss@conference.conversejs.org chat room.

There is another person there, SouL, who is also writing a plugin for converse.js. You two could perhaps help one another with certain aspects of plugin development. I also try to answer questions there when I'm available.

**Comment by happy-dev** [comment]
Sounds excellent. I am coming. 

Which XMPP client do you recommend on Linux (if you use Linux of course) ?

**Comment by jcbrand** [comment]
Gajim is pretty good. I'd recommend that.

I personally use converse.js inside the browser :)

In the repo I run `make serve` and then I open https://localhost:8000/inverse.html (of a modified version of that file for dev purposes).

**Comment by happy-dev** [comment]
Then GO converseJS 
I actually wanted to know if you used something else. 
Converse is more than fine with me

**Comment by jcbrand** [comment]
This has been implemented and will be included in the 4.0.0 release.

**Comment by Heraptor** [comment]
Would like anyone to show how to configure converse.js to enable  "File sharing / HTTP File Upload XEP 363" function ?

**Comment by Pankhur** [comment]
@jcbrand i have setup ejaabberd & converse.js and everything is working fine as per chat is concerned. Now please guide me the way for attachment support in the same. Thanks

**Comment by licaon-kter** [comment]
@Pankhur What error do you see?

**Comment by Pankhur** [comment]
@licaon-kter So, i configured ejabberd.yml config for attachment support as mentioned here 

[https://stackoverflow.com/questions/35578600/mod-http-upload-http-file-upload-xep-0363-using-ruby-on-rails](url)

but nothing seems to working and when i attach any file via converse it just sits idle there.

**Comment by licaon-kter** [comment]
So, error in browser console? Error in ejabberd.log? Actual config of server?

**Comment by pielonet** [comment]
Hi, I configured converse.js 6.0 to work with prosody. I activated XEP-0363 module "http_upload_external" module of prosody and set up a perl file upload server behind nginx (https://github.com/weiss/ngx_http_upload). The paperclip icon did appear and clicking on it allows it to choose a file to upload.
BUT nothing happens then : it fails.
I have the following error in Firefox console : 

`ERROR: <iq xmlns="jabber:client" type="error" to="thierry@example.com/converse.js-103998938" id="3eb64816-63ca-40f5-b50a-68213e4916b4:sendIQ" from="conference.example.com"><error type="cancel"><service-unavailable xmlns="urn:ietf:params:xml:ns:xmpp-stanzas"/></error></iq>`

What is going wrong ?

**Comment by jcbrand** [comment]
Prosody is saying that the service is not available. So you probably misconfigured it.

**Comment by pielonet** [comment]
Issue fixed.
It looks like it was linked to the browser cache not being updated...

**Comment by itsdennian** [comment]
> Issue fixed.
> It looks like it was linked to the browser cache not being updated...

Hi Thierry Kauffmann, can you share how you got it running? I am trying to do the same but it's not working.

**Comment by pielonet** [comment]
> > Issue fixed.
> > It looks like it was linked to the browser cache not being updated...
> 
> Hi Thierry Kauffmann, can you share how you got it running? I am trying to do the same but it's not working.

Hi @itsdennian, It's been a long time (many months) since I did this and I can't remember exactly what I did. What is not working as exepected ?
Kind regards,
